src/persist-client/src/internal/state.rs

```diff
@@ -335,6 +335,65 @@ impl<'a, T> Iterator for HollowBatchRunIter<'a, T> {
     }
 }
 
+impl<T: Timestamp> HollowBatch<T> {
+    pub(crate) fn rewrite_ts(
+        &mut self,
+        frontier: &Antichain<T>,
+        new_upper: Antichain<T>,
+    ) -> Result<(), String> {
+        if PartialOrder::less_equal(&new_upper, frontier) {
```

Comment:
hmm maybe this needs to be `frontier !< new_upper`?

Reply:
Yeah that sounds right! (Equivalent for our timestamp type in practice, but requiring the upper to be strictly greater than the frontier makes it easier to prove that it will never advance any timestamp past the upper.)